### PR TITLE
[Merged by Bors] - feat(RingTheory.MvPolynomial.WeightedHomogeneous): add graded algebra structure

### PR DESCRIPTION
**Pull request by mariainesdff**
Given a weight `w : σ → M`, where `M` is an additive and commutative monoid, we endow the
  ring of multivariate polynomials `MvPolynomial σ R` with the structure of a graded algebra

Co-authored-by: Antoine Chambert-Loir  <antoine.chambert-loir@u-paris.fr>

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/RingTheory/MvPolynomial/Homogeneous.lean [hunk 1]
```diff
@@ -532,13 +533,49 @@ theorem sum_homogeneousComponent :
...
+variable [DecidableEq σ] [DecidableEq R]
+
+/-- The decomposition of `MvPolynomial σ R` into  homogeneous submodules -/
+abbrev decomposition  :
+    DirectSum.Decomposition (homogeneousSubmodule σ R) :=
+  weightedDecomposition R (1 : σ → ℕ)
+
+/-- Given a weight, `MvPolynomial σ R` as a graded algebra -/
+abbrev gradedAlgebra : GradedAlgebra (homogeneousSubmodule σ R) :=
+  weightedGradedAlgebra R (1 : σ → ℕ)
```
Comment (Multramate) [inline Mathlib/RingTheory/MvPolynomial/Homogeneous.lean:563]: This uses `weightedGradedAlgebra`, which has a `DecidableEq R` instance - is this avoidable at all?

Comment (mariainesdff) [reply Mathlib/RingTheory/MvPolynomial/Homogeneous.lean:561]: Good point, thank you. We have removed some instances `DecidableEq R` and `DecidableEq \sigma` which were only used in proofs.

File: Mathlib/RingTheory/GradedAlgebra/Basic.lean [hunk 2]
```diff
@@ -364,3 +364,43 @@ theorem coe_decompose_mul_of_right_mem (n) [Decidable (i ≤ n)] (b_mem : b ∈
...
+
+variable {R : Type*} [CommSemiring R] {A : Type*} [Semiring A] [Algebra R A]
+variable {ι : Type*} [DecidableEq ι] [AddMonoid ι]
+variable {M : ι → Submodule R A} [SetLike.GradedMonoid M]
+
+-- The following lines were given on Zulip by Adam Topaz
+/-- The canonical isomorphism of an internal direct sum with the ambiant algebra -/
+noncomputable def coeAlgIso (hM : DirectSum.IsInternal M) :
+    (DirectSum ι fun i => ↥(M i)) ≃ₐ[R] A :=
+  { RingEquiv.ofBijective (DirectSum.coeAlgHom M) hM with commutes' := fun r => by simp }
```
Comment (jjaassoonn) [inline Mathlib/RingTheory/GradedAlgebra/Basic.lean:379]: Should this be called `coeAlgEquiv`?

File: Mathlib/Algebra/DirectSum/Basic.lean [hunk 3]
```diff
@@ -131,6 +131,18 @@ theorem of_eq_of_ne (i j : ι) (x : β i) (h : i ≠ j) : (of _ i x) j = 0 :=
 lemma of_apply {i : ι} (j : ι) (x : β i) : of β i x j = if h : i = j then Eq.recOn h x else 0 :=
   DFinsupp.single_apply
 
+variable {β} in
```
Comment (semorrison) [inline Mathlib/Algebra/DirectSum/Basic.lean:134]: Probably we just want `variable {β}` starting from before `of_eq_same`. I think lots of these theorems have it too explicitly.

File: Mathlib/Algebra/DirectSum/Basic.lean [hunk 4]
```diff
@@ -398,4 +417,16 @@ theorem IsInternal.addSubmonoid_iSup_eq_top {M : Type*} [DecidableEq ι] [AddCom
   exact Function.Bijective.surjective h
 #align direct_sum.is_internal.add_submonoid_supr_eq_top DirectSum.IsInternal.addSubmonoid_iSup_eq_top
 
+variable  {M S : Type*} [DecidableEq ι] [DecidableEq M] [AddCommMonoid M] [SetLike S M]
+    [AddSubmonoidClass S M] (A : ι → S)
```
Comment (semorrison) [inline Mathlib/Algebra/DirectSum/Basic.lean:421]:
```suggestion
variable {M S : Type*} [DecidableEq ι] [DecidableEq M] [AddCommMonoid M] [SetLike S M]
    [AddSubmonoidClass S M] (A : ι → S)
    
```

File: Mathlib/RingTheory/GradedAlgebra/Basic.lean [hunk 5]
```diff
@@ -364,3 +364,43 @@ theorem coe_decompose_mul_of_right_mem (n) [Decidable (i ≤ n)] (b_mem : b ∈
...
+
+namespace DirectSum.IsInternal
+
+
+variable {R : Type*} [CommSemiring R] {A : Type*} [Semiring A] [Algebra R A]
+variable {ι : Type*} [DecidableEq ι] [AddMonoid ι]
+variable {M : ι → Submodule R A} [SetLike.GradedMonoid M]
+
+-- The following lines were given on Zulip by Adam Topaz
+/-- The canonical isomorphism of an internal direct sum with the ambiant algebra -/
```
Comment (semorrison) [inline Mathlib/RingTheory/GradedAlgebra/Basic.lean:376]:
```suggestion
/-- The canonical isomorphism of an internal direct sum with the ambient algebra -/
```

File: Mathlib/RingTheory/GradedAlgebra/Basic.lean [hunk 6]
```diff
@@ -364,3 +364,43 @@ theorem coe_decompose_mul_of_right_mem (n) [Decidable (i ≤ n)] (b_mem : b ∈
...
+noncomputable def coeAlgIso (hM : DirectSum.IsInternal M) :
+    (DirectSum ι fun i => ↥(M i)) ≃ₐ[R] A :=
+  { RingEquiv.ofBijective (DirectSum.coeAlgHom M) hM with commutes' := fun r => by simp }
+
+/-- Given an `R`-algebra `A` and a family `ι → Submodule R A` of submodules
+parameterized by an additive monoid `ι`
+and statisfying `SetLike.GradedMonoid M` (essentially, is multiplicative)
+such that `DirectSum.is_internal M` (`A` is the direct sum of the `M i`),
```
Comment (semorrison) [inline Mathlib/RingTheory/GradedAlgebra/Basic.lean:384]:
```suggestion
such that `DirectSum.IsInternal M` (`A` is the direct sum of the `M i`),
```

File: Mathlib/RingTheory/GradedAlgebra/Basic.lean [hunk 7]
```diff
@@ -364,3 +364,43 @@ theorem coe_decompose_mul_of_right_mem (n) [Decidable (i ≤ n)] (b_mem : b ∈
...
+we endow `A` with the structure of a graded algebra.
+The submodules are the *homogeneous* parts -/
```
Comment (semorrison) [inline Mathlib/RingTheory/GradedAlgebra/Basic.lean:386]:
```suggestion
The submodules are the *homogeneous* parts. -/
```

File: Mathlib/RingTheory/GradedAlgebra/Basic.lean [hunk 8]
```diff
@@ -364,3 +364,43 @@ theorem coe_decompose_mul_of_right_mem (n) [Decidable (i ≤ n)] (b_mem : b ∈
...
+
+variable {R : Type*} [CommSemiring R] {A : Type*} [Semiring A] [Algebra R A]
+variable {ι : Type*} [DecidableEq ι] [AddMonoid ι]
+variable {M : ι → Submodule R A} [SetLike.GradedMonoid M]
+
+def gradedAlgebra (dM : DirectSum.Decomposition M) : GradedAlgebra M :=
+  { (inferInstance : SetLike.GradedMonoid M) with toDecomposition := dM }
+
+end DirectSum.Decomposition
+-/
```
Comment (semorrison) [inline Mathlib/RingTheory/GradedAlgebra/Basic.lean:406]: Could you either remove, or add an explanatory note why this is present but commented out?

File: Mathlib/RingTheory/MvPolynomial/Homogeneous.lean [hunk 9]
```diff
@@ -532,13 +533,47 @@ theorem sum_homogeneousComponent :
...
 
 end HomogeneousComponent
 
 end
 
+noncomputable section GradedAlgebra
+
+/-- The homogeneous submodules form a graded ring.
+This instance is used by `DirectSum.commSemiring` and `DirectSum.algebra`. -/
+lemma HomogeneousSubmodule.gcomm_monoid :
```
Comment (semorrison) [inline Mathlib/RingTheory/MvPolynomial/Homogeneous.lean:550]: Is this name right? I'm not sure what the principle is here.

File: Mathlib/RingTheory/MvPolynomial/Homogeneous.lean [hunk 10]
```diff
@@ -532,13 +533,47 @@ theorem sum_homogeneousComponent :
...
+    SetLike.GradedMonoid (homogeneousSubmodule σ R) :=
+  WeightedHomogeneousSubmodule.gcomm_monoid
+
+/-- The decomposition of `MvPolynomial σ R` into  homogeneous submodules -/
```
Comment (semorrison) [inline Mathlib/RingTheory/MvPolynomial/Homogeneous.lean:554]:
```suggestion
/-- The decomposition of `MvPolynomial σ R` into homogeneous submodules -/
```

File: Mathlib/RingTheory/MvPolynomial/Homogeneous.lean [hunk 11]
```diff
@@ -532,13 +533,47 @@ theorem sum_homogeneousComponent :
...
+abbrev decomposition  :
```
Comment (semorrison) [inline Mathlib/RingTheory/MvPolynomial/Homogeneous.lean:555]:
```suggestion
abbrev decomposition :
```

File: Mathlib/RingTheory/MvPolynomial/WeightedHomogeneous.lean [hunk 12]
```diff
@@ -313,16 +315,16 @@ theorem weighted_total_degree [SemilatticeSup M] {w : σ → M} (hφ : IsWeighte
     apply Finset.le_sup hd
 #align mv_polynomial.is_weighted_homogeneous.weighted_total_degree MvPolynomial.IsWeightedHomogeneous.weighted_total_degree
 
+end IsWeightedHomogeneous
+
+variable {R}
+
 /-- The weighted homogeneous submodules form a graded monoid. -/
-instance WeightedHomogeneousSubmodule.gcomm_monoid {w : σ → M} :
+lemma WeightedHomogeneousSubmodule.gcomm_monoid {w : σ → M} :
```
Comment (semorrison) [inline Mathlib/RingTheory/MvPolynomial/WeightedHomogeneous.lean:323]: Similarly here, I guess. What is the naming scheme?

File: Mathlib/RingTheory/MvPolynomial/Homogeneous.lean (hunk 9 above)
Comment (mariainesdff) [reply Mathlib/RingTheory/MvPolynomial/Homogeneous.lean:550]: Replaced by `HomogeneousSubmodule.gradedMonoid`.

**Review by riccardobrasca — APPROVED** [review]
Can you add a docstring explaining what to do to get the `GradedAlgebra` instance (with the usual weights) on multivariate polynomials? I mean something like
```
attribute [local instance] MvPolynomial.gradedAlgebra
```
Thanks!

bors d+

**Comment by riccardobrasca** [comment]
bors d=Antoine Chambert-Loir

**Comment by riccardobrasca** [comment]
bors d-antoine

**Comment by riccardobrasca** [comment]
bors d-=antoine

**Comment by riccardobrasca** [comment]
bors d-

**Comment by riccardobrasca** [comment]
bors d=AntoineChambert-Loir

**Pull request closed without merging**